File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        int count = 0; 
        while(root != null) {
            if(root.left == null) {
                if(++count == k)
                    return root.val;
                root = root.right;
            }
            else {
                TreeNode curr = root.left;
                while(curr.right != null && curr.right != root) {
                    curr = curr.right;
                }
                if(curr.right == null) {
                    curr.right = root;
                    root = root.left;
                }
                else {
                    curr.right = null;
                    if(++count == k)
                        return root.val;
                    root = root.right;
                }
            }
        }
        return -1;
    }
}

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getPredecessor(self, node: Optional[TreeNode]) -> Optional[TreeNode]:
        curr = node.left
        while curr.right and curr.right != node:
            curr = curr.right
        return curr

    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        curr = root
        kNode = None
        while curr:
            if curr.left:
                predecessor = self.getPredecessor(curr)
                if predecessor.right == curr:
                    predecessor.right = None
                    k -= 1
                    if k == 0:
                        return curr.val
                    curr = curr.right
                else:
                    predecessor.right = curr
                    curr = curr.left
            else: 
                k -= 1
                if k == 0:
                    return curr.val
                curr = curr.right
        return kNode